### DISCUSSION TITLE
another pure data question

### DISCUSSION THREAD

**Discussion by KatePimbleton** [pschatzmann/arduino-audio-tools#1447] [Q&A] [answered]
Hello,

I'm using arduino-audio-tools ( cloned with git )
and an esp32-audio-kit

i'm trying to merge 2 mono FileLoops, copy the stereo stream to puredata, and output from pd to I2S.
It seems to work, kind of, the sound stutters ( a bit in the same way as when LOG is left to Info or Debug )
I tried to use a bigger buffer for stream copy but if it is set to 1024 my audiokit keeps rebooting.

the puredata patch used is the same as in the stream example, 2 in 2 outs, nothing else.

Am I exceeding the audiokit or sdcard capabilities ?

here's the code I came up with:

```C++
#include <SPI.h>
#include <SD.h>
#include "AudioTools.h"
#include "AudioLibs/AudioBoardStream.h"
#include "AudioLibs/FileLoop.h"
#include "Heavy_fileLoopPd.hpp"
#include "AudioLibs/PureDataStream.h"

const int chipSelect = PIN_AUDIO_KIT_SD_CARD_CS;

AudioInfo info(44100, 2, 16);
InputMerge<int16_t> merge;
Heavy_fileLoopPd pd_test(44100);
PureDataStream pd(pd_test);

AudioBoardStream i2s(AudioKitEs8388V1);  // final output of decoded stream
FileLoop loopingFile;
FileLoop loopingFile2;
EncodedAudioStream decoder(&loopingFile, new WAVDecoder());
EncodedAudioStream decoder2(&loopingFile2, new WAVDecoder());
StreamCopy copierToPd(pd, merge,256);
StreamCopy copierFromPd(i2s, pd,256);


void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);
  auto config = i2s.defaultConfig(TX_MODE);
  config.copyFrom(info);
  config.sd_active = true;
  i2s.begin(config);
  SD.begin(chipSelect);
  loopingFile.setFile(SD.open("/seanMono.wav"));
  loopingFile2.setFile(SD.open("/pianoMono.wav"));
  loopingFile.begin();
  loopingFile2.begin();
  merge.add(loopingFile);
  merge.add(loopingFile2);
  merge.begin(info);
  decoder.begin();
  decoder2.begin();
  pd.begin();
}
void loop() {
  copierToPd.copy();
  copierFromPd.copy();
}
```

**Comment by pschatzmann** [comment, marked as answer]
Difficult to say: I suggest that you try to test the different pieces separately before you integrate them into PD: e.g. input only & copy to i2s, input merge & copy to I2S

A couple of additional thoughts:
- you can try to lower the sample rate
- try to upgrade the AudioTools library: I don't know if you have the FileLoop implementation already which removes the WAV header automatically when moving back to the beginning. You can compare your local version with the [actual implementation](https://github.com/pschatzmann/arduino-audio-tools/blob/main/src/AudioLibs/FileLoop.h). Or call setStartPos(44) to ignore the header.

**Reply by KatePimbleton** [reply]
Cheers, I lowered the sample rate by half (22050hz), and it works fine.
I also tried bypassing Puredata, it works fine at 44100hz.
Maybe I could increase buffers, I did it in AudioConfig.h but 
without exactly knowing what I do and with no luck.
How do I increase the buffers ?

**Reply by pschatzmann** [reply]
- Make sure that the log level is set to warning.
- In the constructor of the PureDataStream you can indicate the max buffer size: Currently this is 2K which is a bit too high and should better match with the copy size.
- In the constructor of the copier you can indicate how many bytes should be copied with each loop step. Try to increase the number. I think you might then also need to increase the stack size.
- The other place where you could play with is the i2s configuration in the sketch by setting the cfg.buffer_size and cfg.buffer_count.

**Reply by KatePimbleton** [reply]
cheers!

**Comment by pschatzmann** [comment]
I [changed](https://github.com/pschatzmann/arduino-audio-tools/commit/9b5930ea671bbc7face8ec401aca7dadf98f835d) the PureDataStream to avoid any arrays on the stack. With the updated version it should be possible to increase the copy size e.g. to 1024
